### ISSUE TITLE
[DISCUSSION] YARN 1.0.0 release

### ISSUE THREAD

**Issue by kevin-bates** [toidi/hadoop-yarn-api-python-client#47]
Per our 1.0.0 release discussion on #34 and talking a bit with @lresende, we think it might be best to jump directly to `1.0.0` and by-pass a pre-release of `1.0.0b1` or `1.0.0xxx`.  The reason for this is two-fold:
1. pip installs would require `--pre` to access the pre-release.
2. Dependent applications cannot add `yarn-api-client>=1.0.0` and have that pull down `1.0.0b1` when installing the dependent application.

As a result, it might be best to just go to `1.0.0` which enables dependent applications to pin a minimum version and pip installs "just work".

We would then use `1.0.x` for any immediate bugs - essentially becoming a "post-release beta". :smile:

This also adheres to the mantra that API-breaks require an increment to major version (as we've discussed), so using something like `0.9.x` as our "pre-release" carriers is also a violation.

Summary: Proposal is to optimistically move to `1.0.0` while creating maintenance releases to stabilize.

Comments, suggestions?

**Comment by dimon222** [comment]
Sadly, until its tested on real environment I don't see point marking 1.0.0 yet.
I understand that there are certain complexities with installation, but they're there for reason - making sure people don't randomly deploy it for production assuming "it works" 💯 

Before we reach 1.0.0 I think we might want to close next gaps:
1. Validate that none of current API endpoints are not matching to latest release of Hadoop YARN API spec (not missing endpoints, but instead - changed/deleted endpoints). If everything is confirmed matching - put timestamp with version release of hadoop on which it validated - to track new changes of API spec from this moment and give "somewhat" contract for potential users that "well, it works on Hadoop 3.0.0 because it was developed for this version spec".
2. If possible, cover missing REST endpoints at least for STABLE api spec (and optionally - for ALPHA once STABLE is covered)
3. Spend some time walking through code and confirm that some "to do" comment gaps are covered properly. I believe there are not many, but why not?
4. We may want to add some examples of using kerberized environment (with login, etc). Currently we have automatically generated documentation based of codebase, but I'm sure there will be many users who would ask "i dont understand, how to login". Example in my https://github.com/toidi/hadoop-yarn-api-python-client/pull/34#issue-283485863 would be useful in some documentation page "how to cook this"
5. There are also certain things I noticed that make `pytest` impossible to completely run on non-hadoop local environment because it expects `yarn-site.xml`. Possibly, might require some tuning to make tests simple across different machines (we want #18 #17  anyway, right?)

**Comment by kevin-bates** [comment]
@dimon222 - these seem like reasonable items - thank you for bringing this up.  This is more "correct".

So others have _some_ means of installing a 1.0 pre-release, I think `1.0.0b1` makes sense.  This will be tied to `master` and we can create one or more release candidates as things stabilize.

Should an issue come up against `0.3.7`, we'll need to create a `0.x` branch in which to apply those kinds fixes and build `0.x.y` against that branch.

How does that approach sound?

**Comment by dimon222** [comment]
@kevin-bates sure, I think porting some endpoint fixes shouldn't be problem.

One more radical approach for testing reasons:
`pip install git+https://github.com/toidi/hadoop-yarn-api-python-client.git`

**Comment by kevin-bates** [comment]
I'll look at getting a `1.0.0b1` release out tomorrow - unless future comments preempt.

**Comment by dimon222** [comment]
Tl;dr progress so far for Hadoop 3.2.0 coverage

- [x] ResourceManager (pending in #48 )
- [x] NodeManager (pending in #49 for minor cleanup)
- [ ] ApplicationManager
- [ ] HistoryServer
- [ ] [Timeline Service v1](https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/TimelineServer.html#Timeline_Server_REST_API_v1) 
- [ ] [Timeline Service v2](https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/TimelineServiceV2.html#Timeline_Service_v.2_REST_API)

AM and HS are very specific to MapReduce, so probably "indirect" API of yarn, so I'm not sure if we should really rush those as its way too many APIs anyway.
Timeline Service is altogether brand new components will be needed (since two different api version)+ considering amount of effort required for that - we probably want to mention in README that this is not yet part of implemented components, and can be target for some another major release after 1.x is out

**Comment by lresende** [comment]
What is your targeted audience here for the beta release? A beta release is considered a pre-release and I will not get installed by default. Another approach is to create a 0.4 or 0.9 release and enable consumers and upstream projects to easily test and validate the recent changes while we finalize the open items listed above.

**Comment by dimon222** [comment]
@lresende not sure if directed to me or to @kevin-bates 
For me specifically no really big point in beta, I believe with recent PRs I've prepared most of missing parts so should be pretty close to full pledged test stage.

Our debt - integration tests, Timeline Service APIs, some documentation - but this can be added going on.

**Comment by kevin-bates** [comment]
@dimon222 - thanks for the great work getting the new APIs in shape.  I agree that other things are less pressing.

@lresende - I believe a 1.0 release is better (and more community-friendly) than say, 0.4 or 0.9 simply because it won't _force_ the community to change their code.  Even though the same could be said once 1.0 is available, it's a major release and those kinds of things, while inconvenient, are somewhat expected.  The same is not true for 0.4/0.9 other than they're "protected" by the "you can do anything within a zero major release".

I'm going to build a version and take it for a spin on Enterprise Gateway - although we use exactly 9 methods - all from `ResourceManager`- so it's not a comprehensive test.  That said, I'd like to get an idea for how close we are to cutting 1.0?

**Comment by dimon222** [comment]
I've validated changelog for Hadoop 3.2.1 that got released literally two days ago, but API changes we applied are still all up to date (there were changes only for Timeline Service v2 but we don't have it implemented)

**Comment by kevin-bates** [comment]
That's great news Dmitry - thanks for your great work.  

Are we ready to cut 1.0? 

cc: @lresende, @toidi, @dimon222

**Comment by dimon222** [comment]
We can probably close this, since release already happened.

**Comment by kevin-bates** [comment]
Right on - thanks for the reminder!
